elastic-buffer-sim/exe/Main.hs

```diff
@@ -2,70 +2,644 @@
+        Left err        -> die $ "ERROR: Invalid DOT file - " <> f <> "\n" <> err
+    Nothing ->
+      handleParseResult $ Failure
+        $ parserFailure defaultPrefs cliParser (ShowHelpText Nothing) []
+
+  let topologyFile = outDir </> "topology.gv"
+  writeFile topologyFile $ (<> "\n") $ render $ toDot g name
+  BS.writeFile (outDir </> "simulate.json") $ encode options
+     { stable   = isStable
+     , offsets  = offs
+     , jsonArgs = Nothing
+     , topology = case topology of
+         Just (Random _) -> Just $ DotFile topologyFile
+         _               -> topology
+     }
+
+  if isStable
+  then exitSuccess
+  else exitFailure
+
+randomGraph :: Int -> IO Graph
+randomGraph n = do
+  let is = [0,1..n - 1]
+
+  -- get some random vertex permuation for ensuring connectivity
+  aP <- newListArray (0, n - 1) is
+  replicateM_ (10 * n) $ do
+    p1 <- randomRIO (0, n - 1)
+    p2 <- randomRIO (0, n - 1)
+    v1 <- readArray aP p1
+    v2 <- readArray aP p2
+    writeArray aP p1 v2
+    writeArray aP p2 v1
+
+  randomPermutation <- getElems (aP :: IOUArray Int Int)
+
+  -- get some random edge availability matrix for self-loop-free,
+  -- undirected graphs
+  aE <- replicateM (n * n) randomIO >>= newListArray ((0, 0), (n - 1, n - 1))
+  forM_ is $ \i ->
+    writeArray aE (i, i) False
+  forM_ [(i, j) | i <- is, j <- is, i /= j] $ \(i, j) -> do
+    d1 <- readArray aE (i, j)
+    d2 <- readArray aE (j, i)
+    when (d1 /= d2) $ do
+      writeArray aE (i, j) False
+      writeArray aE (j, i) False
+
+  -- ensure connectivity
+  forM_ (unchain randomPermutation) $ \(i, j) -> do
+    writeArray aE (i, j) True
+    writeArray aE (j, i) True
+
+  available <- freeze (aE :: IOUArray (Int, Int) Bool)
+
+  -- create the graph
+  return $ buildG (0, n - 1)
+    [ (i, j)
+    | i <- is
+    , j <- is
+    , available A.! (i, j)
+    ]
+
+toDot :: Graph -> String -> (Bool, GraphType, Maybe Name, [Statement])
+toDot g name = (True, Graph, Just $ StringID name, map asEdgeStatement $ edges g)
+ where
+  asEdgeStatement (x, y) =
+    EdgeStatement (map (\i -> NodeRef (XMLID ('n' : show i)) Nothing) [x,y]) []
+
+fromDot :: String -> Either String (Graph, String)
+fromDot cnt = do
+  (strict, gType, maybe "" fromDotName -> name, statements) <- parse cnt
+
+  when (not strict) $ Left "Graph must be strict"
+  when (gType == Digraph) $ Left "Graph must be undirected"
+
+  namedEdges <-
+    fmap
+      ( concatMap (map (\(x,y) -> (asString x, asString y)) . unchain)
+      . catMaybes
+      )
+    $ forM statements $ \case
+        EdgeStatement xs _ -> fmap Just $ forM xs $ \case
+          NodeRef n _ -> return n
+          Subgraph{}  -> Left "Subgraphs are not supported"
+        _ -> pure Nothing
+
+  let
+    edgeNames = strictSort $ concatMap (\(x, y) -> [x, y]) namedEdges
+    n = length edgeNames - 1
+    idx = (M.!) $ M.fromList $ zip edgeNames [0,1..]
+    g = buildG (0,n)
+      $ filter (uncurry (/=))
```

Comment:
Does it make sense to silently drop self-loops? Isn't this an input error?

Reply:
That depends on what you consider the selfloop to express in terms of a bittide topology. In terms of connectivity, a node can synchronize its clock with itself even without a elastic buffer in between. Therefore, for me graphs with self-loops as input are just fine. The self-loops just don't have any effect on the topology that is created out of it.

Reply:
That makes sense, add that as a comment.

Reply:
Done.